### ISSUE TITLE
Curriculum mindset homework

### ISSUE THREAD

**Issue by tkphd** [hpc-carpentry/coordination#46]
For our next Coordination Meeting (Thursday, May 27), after elections discussions, I would like to focus on curriculum development and the "scope" of our existing and planned lessons. To that end, I would recommend the online book @ocaisa  recently mentioned, http://teachtogether.tech/en/ by Greg Wilson (Carpentries guru *par excellence*). I can't actually assign homework, but if you have time over the next month, please spend some of it reading. As a teaser, here are "The Rules" Greg recommends we all keep in mind:

1. Be kind: all else is details.
2. Remember that you are not your learners…
3. …that most people would rather fail than change…
4. …and that ninety percent of magic consists of knowing one extra thing.
5. Never teach alone.
6. Never hesitate to sacrifice truth for clarity.
7. Make every mistake a lesson.
8. Remember that no lesson survives first contact with learners…
9. …that every lesson is too short for the teacher and too long for the learner…
10. …and that nobody will be more excited about the lesson than you are.

**Comment by bkmgit** [comment]
Co-working meeting raised issue of using Software carpentry shell lesson, and augmenting this material to minimize repeated work required for maintaining [HPC Shell](http://hpc-carpentry.org/hpc-shell/).  There are a number of shell lessons:
a) [Library Carpentry](https://librarycarpentry.org/lc-shell/)
b) [Software Carpentry](http://swcarpentry.github.io/shell-novice/)
c) [Shell Genomics](https://datacarpentry.org/shell-genomics/)
d) [Shell economics](https://datacarpentry.org/shell-economics)

Since all of these are likely users of HPC resources, one should only presume material contained in all of these shell lessons, so that any one of them can fulfill prerequisite for HPC intro.

**Comment by reid-a** [comment]
Meeting date is wrong in the original issue, this meeting is scheduled for May 20, 2021.

**Comment by bkmgit** [comment]
The potential learners need to be considered. 
 - The current HPC intro lesson is for people who are doing basic programming already and need more computational resources. A minimum of a Python or Julia background is expected here. The main outcome is a conceptual understanding of a parallel program as well as being able to submit a job.
 - People with no programming background from non traditional fields who will need to learn both programming and parallelization. Many of these will likely be motivated by big data problems. The Chapel lesson is good for this, though the examples maybe a little abstract for applied users, but maybe great for younger users interested in skills and concepts. [DeapSecure](https://deapsecure.gitlab.io/) has good motivating big data examples.
 - People who use HPC because their programs (which they have written or inherited) run slowly, and will need an introduction to performance engineering. This is a possible further lesson.

**Comment by aturner-epcc** [comment]
@bkmgit I disagree with the statement

> The current HPC intro lesson is for people who are doing basic programming already and need more computational resources. A minimum of a Python or Julia background is expected here. The main outcome is a conceptual understanding of a parallel program as well as being able to submit a job.

The current hpc-intro is an introduction to the use of remote HPC facilities for any researcher who wants to use HPC. Some of these may be doing basic programming but many of them will be using software packages that are already installed on the system and use hpc-intro to learn how to

* get a basic concept of what makes up an HPC system and how it fits together
* connect to and interact with the system
* manage the software environment on the system
* write job scripts and submit jobs
* understand what using a shared system means for them and their data

We use the lesson as our gateway training for users who have never used HPC before and have zero programming experience and it works really well for that cohort.

I think your list of potential learners misses this type of person:

* Researchers who need to use HPC for their research (as the computational problem is intractable on local resources) but are not developing their own parallel software. Rather, they are using preinstalled parallel software on the HPC system.

**Comment by bkmgit** [comment]
@aturner-epcc Great. It seems we have some good personas. Do they have a linux/shell introduction before hpc-intro? Should the amount of programming in HPC intro be reduced or the example changed? The current lesson motivates parallel computing as providing faster time to solution and/or more memory. The more memory part is not well motivated by the current pi calculation example since one can argue that this can be more efficiently programmed. Alternatively, maybe one could use this as an exercise to discuss ways of evaluating effective use of computational resources?

**Comment by aturner-epcc** [comment]
@bkmgit We generally run an optional hpc-shell session for people before hpc-intro. If people already have some command line experience they can skip straight to hpc-intro.

We modify the "Scheduling Jobs" episode to cover launching parallel jobs with an MPI image sharpening example and add some more work to the "Using Resources Effectively" episode to allow them to do some basic benchmarking. We drop the "Running a parallel job" episode completely as it is not useful in the context of the learners we have for this course. See, for example:

* https://epcced.github.io/hpc-intro-2021-02-20/13-scheduler/index.html
* https://epcced.github.io/hpc-intro-2021-02-20/16-resources/index.html

These make it more suitable for non-programmers.

If people want to learn parallel programming, we have different courses for that:

* https://github.com/EPCCed/archer2-MPI-2021-03-17
* https://www.archer2.ac.uk/training/courses/201006-openmp/
* https://github.com/EPCCed/archer-gpu-course

**Comment by aturner-epcc** [comment]
I think that dropping the "Running a parallel job" episode from "hpc-intro" and using "parallel-novice" to teach basic parallel programming probably makes more sense from a learner perspective.

**Comment by tkphd** [comment]
@reida commented during today's Coworking Hour that teaching the mechanics of launching a job on the cluster is not enough: the Monte Carlo pi estimator code is *value added* because it introduces MPI concepts in a concrete and useful example. Handing learners a packaged program to just launch on the cluster would exercise the job submission workflow, but with limited value to learners who are building a mental model of parallel computing as we go.

**Comment by aturner-epcc** [comment]
Certainly, gaining a mental model of parallel computation and the potential impacts on performance (why is my parallel calculation not faster than the serial one or one on less cores) is important. However, this is very different from having a model of how MPI works - which is specifically important if you are going to be programming in MPI.

The precompiled sharpen example we use does not introduce MPI concepts (other than launching a parallel MPI code) but it does introduce learners to assessing scaling properties and to the concepts of Amdahl's Law. This is actually important and useful for them in building a model of parallel computation that is useful to them even though they are not developing their parallel programs. Furthermore, it is *generally* useful in having a model of parallel computing rather than the specific model of MPI - as it also provides a basis for basic understanding in all different parallel models: shared memory (e.g. OpenMP), SIMD (e.g. AVX, CUDA, ROCm), RDMA (CAF, SHMEM), task based parallelism (e.g. dask, gnu parallel, snakemake).

In my opinion, the PI MPI example adds a lot of additional cognitive load: understanding MPI concepts, which are of little benefit to many of the learners (who will jot be going on to learn MPI). And, for those learners that do go on to learn MPI it is not really enough to give them a useful foundation - they need a full intro to MPI course for that.

**Comment by tkphd** [comment]
Excellent comment, @aturner-epcc, thank you.

**Comment by bkmgit** [comment]
@aturner-epcc The organization of your lessons seems ok.  It is helpful to be able to see parallel source code, and not only run a pre-compiled binary, even though this may be the end goal for some users.  Modification of the lesson to use software already installed as a module on the particular cluster is a good option for this.  A flexible lesson should not assume much in terms of pre-installed software though. Instructor can make a teaching choice to decide how much depth to go into on programming aspect since not all locations will have the time to do a more extensive parallel programming course. The current pi example does not effectively motivate memory use which maybe one reason to use HPC and neither does sharpen, maybe the lesson is ok without this, though one could have an exercise asking for amount of memory per node and then do a calculation for number of nodes needed to obtain say 2 Tb.  Possibly one can modify/augment the warm up introductory ice breaker exercise to go through the steps of a parallel algorithm (calculating pi or sorting or searching or word count) and then have implementations of this in a number of languages.

**Comment by bkmgit** [comment]
Part of Carpentries philosophy is creating a growth mindset.  Having some kind of source code better enables a growth mindset, as indicated in http://teachtogether.tech/en/#s:intro-exercises 

"Politicians, business leaders, and educators often say that people should learn to program because the jobs of the future will require it. However, as Benjamin Doxtdator [pointed out](http://longviewoneducation.org/field-guide-jobs-dont-exist-yet/), many of those claims are built on shaky ground. Even if they were true, education shouldn’t prepare people for the jobs of the future: it should give them the power to decide what kinds of jobs there are and to ensure that those jobs are worth doing. And as Mark Guzdial [points out](https://computinged.wordpress.com/2017/10/18/why-should-we-teach-programming-hint-its-not-to-learn-problem-solving/), there are actually many reasons to learn how to program:
- To understand our world.
- To study and understand processes.
- To be able to ask questions about the influences on their lives.
- To use an important new form of literacy.
- To have a new way to learn art, music, science, and mathematics.
- As a job skill.
- To use computers better.
- As a medium in which to learn problem-solving."

**Comment by aturner-epcc** [comment]
@bkmgit I agree with the principles that you state here and think you are right that we should not presume exactly what people would want to teach and that we need to have flexibility here.

How about this for a proposal for a way forward for the lesson here:
* Push the sharpen version of the resources episode example into the extras (along with the source code: there are Fortran MPI, OpenMP, serial and C MPI, OpenMP, serial versions).
* Add instructor notes about the options for choosing to use the pi example or sharpen example along with the lesson layout that would enable this and what the differences mean for different audiences
* Add instructor notes on the steps required to switch from the default (pi) example to the sharpen example

I am happy to start putting together a PR that would enable this with  sharpen code and example.

As for your other points:
* Memory use: one challenge here is that even people who regularly develop HPC codes are not as aware of their memory requirements. While it is easy to envisage working out how much memory a number of nodes gives you I think it is much harder in general for people (novices and experts) to have a good a priori estimate of how much memory their application (program + input case) will require which makes this difficult to capture in a useful exercise.
* Parallel algorithms: this is an interesting idea. I wonder if a good way to approach this is to give some common parallel decompositions with concrete, real world examples to help build a good mental model of how applications actually exploit HPC systems (a bit like the bullet points at the start of episode 11 that motivate the use of HPC but focussed on algorithms). For example:
   * Trivial parallelism: pi, mandlebrot set
   * Grids: CFD, climate modelling
   * Domain decomposition, N-body: biomolecular modelling, cosmology
   * Dense linear algebra: electronic structure materials modelling

What do you think?

**Comment by bkmgit** [comment]
This seems good.

Sharpen example is nice. It may also fit well into the novice lesson. Similarly, I think the algorithms would also fit well into the novice lesson, but can be introduced as application areas of HPC in the intro lesson with representative images and links to further reading for those who want it.  One or two algorithms is probably sufficient for non-programmers.

For memory, there are nuances on what the runtime, libraries and operating system will do, so this needs to be done carefully. One of the motivating points for using HPC resources is to solve bigger problems. Still in favor of a calculation exercise with a comment, since as you indicate some experimentation is required

There is a parallel sort demo at  https://github.com/EPCCed/Activities/tree/master/Activities as well as cellular automata and addition examples at https://www.futurelearn.com/courses/supercomputing - cellular automata are very similar to sharpen so this can also be motivated by a thought/discussion experiment at the beginning of the lesson.

**Comment by bkmgit** [comment]
For the intro lesson, probably the distributed memory versions are sufficient.
